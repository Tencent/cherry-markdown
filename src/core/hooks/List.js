/**
 * Copyright (C) 2021 THL A29 Limited, a Tencent company.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import ParagraphBase from '@/core/ParagraphBase';

const INDENT_SPACE_NUM = 4; // commonmark default use 1~4 spaces for indent
const TAB_SPACE_NUM = 4; // 1 tab === 4 space

function attrsToAttributeString(object) {
  if (typeof object !== 'object' && Object.keys(object).length < 1) {
    return '';
  }
  const attrs = ['']; // 为了join一步到位
  Object.keys(object).forEach((key) => {
    attrs.push(`${key}="${object[key]}"`);
  });
  return attrs.join(' ');
}

export function makeChecklist(text) {
  return text.replace(/^((?:|[\t ]+)[*+-]\s+)\[(\s|x)\]/gm, (whole, pre, test) => {
    const checkHtml = /\s/.test(test)
      ? '<span class="ch-icon ch-icon-square"></span>'
      : '<span class="ch-icon ch-icon-check"></span>';
    return `${pre}${checkHtml}`;
  });
}

// 缩进处理
function handleIndent(str, node) {
  const indentRegex = /^(\t|[ ])/;
  let $str = str;
  while (indentRegex.test($str)) {
    node.space += $str[0] === '\t' ? TAB_SPACE_NUM : 1;
    $str = $str.replace(indentRegex, '');
  }
  return $str;
}

// 序号样式处理
function getListStyle(m2) {
  if (/^[a-z]/.test(m2)) {
    return 'lower-greek';
  }
  if (/^[一二三四五六七八九十]/.test(m2)) {
    return 'cjk-ideographic';
  }
  if (/^I/.test(m2)) {
    return 'upper-roman';
  }
  if (/^\+/.test(m2)) {
    return 'circle';
  }
  if (/^\*/.test(m2)) {
    return 'square';
  }
  return 'default';
}

// 标识符处理
function handleMark(str, node) {
  const listRegex =
    /^((([*+-]|\d+[.]|[a-z]\.|[I一二三四五六七八九十]+\.)[ \t]+)([^\r]*?)($|\n{2,}(?=\S)(?![ \t]*(?:[*+-]|\d+[.]|[a-z]\.|[I一二三四五六七八九十]+\.)[ \t]+)))/;
  if (!listRegex.test(str)) {
    node.type = 'blank';
    return str;
  }
  return str.replace(listRegex, (wholeMatch, m1, m2, m3, m4) => {
    node.type = m2.search(/[*+-]/g) > -1 ? 'ul' : 'ol';
    node.listStyle = getListStyle(m2);
    node.start = Number(m2.replace('.', '')) ? Number(m2.replace('.', '')) : 1;
    return m4;
  });
}

class Node {
  // 列表树节点
  constructor() {
    this.index = 0;
    this.space = 0;
    this.type = '';
    this.start = 1;
    this.listStyle = '';
    this.strs = [];
    this.children = [];
    this.lines = 0;
  }
}

export default class List extends ParagraphBase {
  static HOOK_NAME = 'list';

  constructor({ config }) {
    super({ needCache: true });
    this.config = config || {};
    this.tree = [];
    this.emptyLines = 0;
    this.indentSpace = Math.max(this.config.indentSpace, 2);
  }

  addNode(node, current, parent, last) {
    if (node.type === 'blank') {
      this.tree[last].strs.push(node.strs[0]);
    } else {
      this.tree[parent].children.push(current);
      this.tree[current] = {
        ...node,
        parent,
      };
    }
  }

  buildTree(html, sentenceMakeFunc) {
    const items = html.split('\n');
    this.tree = [];
    items.unshift('');
    // 列表结尾换行符个数
    const endLineFlagLength = html.match(/\n*$/g)[0].length;
    for (let i = 0; i < items.length - endLineFlagLength; i++) {
      const node = new Node();
      items[i] = handleIndent(items[i], node);
      items[i] = handleMark(items[i], node);
      node.strs.push(sentenceMakeFunc(items[i]).html);
      node.index = i;
      if (i === 0) {
        // 根节点
        node.space = -2;
        this.tree.push(node);
        continue;
      }
      let last = i - 1;
      while (!this.tree[last]) last -= 1;
      if (node.type === 'blank') {
        this.addNode(node, i, this.tree[last].parent, last);
      } else {
        while (!this.tree[last] || this.tree[last].space > node.space) last -= 1;
        const { space } = node;
        const lastSpace = this.tree[last].space;
        if (space < lastSpace + this.indentSpace) {
          // 成为同级节点
          if (this.config.listNested && this.tree[last].type !== node.type) {
            this.addNode(node, i, last);
          } else {
            this.addNode(node, i, this.tree[last].parent);
          }
        } else if (space < lastSpace + this.indentSpace + INDENT_SPACE_NUM) {
          // 成为子节点
          this.addNode(node, i, last);
        } else {
          // 纯文本
          node.type = 'blank';
          this.addNode(node, i, this.tree[last].parent, last);
        }
      }
    }
  }

  renderSubTree(node, children, type) {
    let lines = 0;
    const attr = {};
    const content = children.reduce((html, item) => {
      const child = this.tree[item];
      const itemAttr = {};
      const str = `<p>${child.strs.join('<br>')}</p>`;
      child.lines += this.getLineCount(child.strs.join('\n'));
      const children = child.children.length ? this.renderTree(item) : '';
      node.lines += child.lines;
      lines += child.lines;
      // checklist 样式适配
      const checklistRegex = /<span class="ch-icon ch-icon-(square|check)"><\/span>/;
      if (checklistRegex.test(str)) {
        itemAttr.class = 'check-list-item';
      }
      return `${html}<li${attrsToAttributeString(itemAttr)}>${str}${children}</li>`;
    }, '');
    if (node.parent === undefined) {
      // 根节点增加属性
      attr['data-lines'] = node.index === 0 ? lines + this.emptyLines : lines;
      attr['data-sign'] = this.sign;
    }
    if (children[0] && type === 'ol') {
      attr.start = this.tree[children[0]].start;
    }
    attr.class = `cherry-list__${this.tree[children[0]].listStyle}`;
    return `<${type}${attrsToAttributeString(attr)}>${content}</${type}>`;
  }

  renderTree(current) {
    let from = 0;
    const node = this.tree[current];
    const { children } = node;
    const html = children.reduce((html, item, index) => {
      if (index === 0) return html;
      if (this.tree[children[index]].type === this.tree[children[index - 1]].type) {
        return html;
      }
      const subTree = this.renderSubTree(node, children.slice(from, index), this.tree[children[index - 1]].type);
      from = index;
      return html + subTree;
    }, '');

    const childrenHtml = children.length
      ? this.renderSubTree(node, children.slice(from, children.length), this.tree[children[children.length - 1]].type)
      : '';

    return html + childrenHtml;
  }

  toHtml(wholeMatch, sentenceMakeFunc) {
    // 行数计算吸收的空行
    this.emptyLines = wholeMatch.match(/^\n\n/)?.length ?? 0;
    const text = wholeMatch.replace(/~0$/g, '').replace(/^\n+/, '');
    this.buildTree(makeChecklist(text), sentenceMakeFunc);
    const result = this.renderTree(0);
    return this.pushCache(result, this.sign, this.$getLineNum(wholeMatch));
  }

  $getLineNum(str) {
    const beginLine = str.match(/^\n\n/)?.length ?? 0;
    const $str = str.replace(/^\n+/, '').replace(/\n+$/, '\n');
    return $str.match(/\n/g)?.length ?? 0 + beginLine;
  }

  makeHtml(str, sentenceMakeFunc) {
    let $str = `${str}~0`;
    if (this.test($str)) {
      $str = $str.replace(this.RULE.reg, (wholeMatch) => {
        return this.getCacheWithSpace(
          this.checkCache(wholeMatch, sentenceMakeFunc, this.$getLineNum(wholeMatch)),
          wholeMatch,
        );
      });
    }
    $str = $str.replace(/~0$/g, '');
    return $str;
  }

  rule() {
    const ret = {
      begin: '(?:^|\n)(\n*)(([ ]{0,3}([*+-]|\\d+[.]|[a-z]\\.|[I一二三四五六七八九十]+\\.)[ \\t]+)',
      content: '([^\\r]+?)',
      end: '(~0|\\n{2,}(?=\\S)(?![ \\t]*(?:[*+-]|\\d+[.]|[a-z]\\.|[I一二三四五六七八九十]+\\.)[ \\t]+)))',
    };
    ret.reg = new RegExp(ret.begin + ret.content + ret.end, 'gm');
    return ret;
  }
}
