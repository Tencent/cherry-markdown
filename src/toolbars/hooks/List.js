/**
 * Copyright (C) 2021 THL A29 Limited, a Tencent company.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import MenuBase from '@/toolbars/MenuBase';
import { getSelection } from '@/utils/selection';
import { getListFromStr } from '@/utils/regexp';
/**
 * 插入有序/无序/checklist列表的按钮
 */
export default class List extends MenuBase {
  constructor($cherry) {
    super($cherry);
    this.setName('list', 'list');
    this.subMenuConfig = [
      { iconName: 'ol', name: 'ol', onclick: this.bindSubClick.bind(this, '1') },
      { iconName: 'ul', name: 'ul', onclick: this.bindSubClick.bind(this, '2') },
      { iconName: 'checklist', name: 'checklist', onclick: this.bindSubClick.bind(this, '3') },
    ];
  }

  getSubMenuConfig() {
    return this.subMenuConfig;
  }

  /**
   * 响应点击事件
   * @param {string} selection 编辑区选中的文本内容
   * @param {1|2|3|'ol'|'1'|'2'|'3'|'ul'|'checklist'|''} shortKey 快捷键：ol(1)有序列表，ul(2)无序列表，checklist(3) 检查项
   * @returns 对应markdown的源码
   */
  onClick(selection, shortKey = '') {
    const listType = [null, 'ol', 'ul', 'checklist']; // 下标1, 2, 3生效
    const $selection = getSelection(this.editor.editor, selection, 'line', true);
    const [before] = $selection.match(/^\n*/);
    const [after] = $selection.match(/\n*$/);
    const type = listType[shortKey] ? listType[shortKey] : shortKey;

    if (!type || !/^(ol|ul|checklist)$/.test(type)) {
      return $selection;
    }
    return `${before}${getListFromStr($selection, type)}${after}`;
  }
}
