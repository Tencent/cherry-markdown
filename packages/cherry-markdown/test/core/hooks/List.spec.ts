import List from '../../../src/core/hooks/List';
import md5 from 'md5';

const cases = []
cases[0] = `
- 1
- 2
  - 2.1
  - 2.2
- 3
  + 3.1
- 4
  * 4.2
`;

cases[1] = `
- 1
  - 2
       - 2.1
    - 2.2
 - 3
          + 3.1
 - 4
* 4.2
`;

cases[2] = `
- 1
1. test

- 1.1
   - 1.1.2
       - blank
  - 1.2
 - 2
          + blank
      - 2.1
	* 2.2
 1. test
   2. 2
`;

cases[3] = `
1. test
	2. test
1. test
   ä¸€. test
   1. test
   
   
   a. test
- test
`;

cases[4] = `
- [ ] checklist 1
- test
  - [x] checklist 2
 - [ ] checklist 3
 - test
      - [ ] checklist 4
`;

const listHook = new List({ config: {
  indentSpace: 2,
}});

Object.defineProperty(listHook, '$engine', {
  value: { md5 },
});

describe('core/hooks/list', () => {
  it('list hook', () => {
    cases.forEach((item) => {
      listHook.makeHtml(item, (text) => ({ html: text }));
      expect(listHook.cache[listHook.sign].content).toMatchSnapshot();
    });
  });
});
