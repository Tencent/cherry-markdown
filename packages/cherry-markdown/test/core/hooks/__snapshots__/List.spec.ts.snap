// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`core/hooks/list list hook 1`] = `"<ul data-lines=\\"8\\" data-sign=\\"2052aa9811ccbbdfcace4f3e56d89249\\" class=\\"cherry-list__default\\"><li><p>1</p></li><li><p>2</p><ul class=\\"cherry-list__default\\"><li><p>2.1</p></li><li><p>2.2</p></li></ul></li><li><p>3</p><ul class=\\"cherry-list__circle\\"><li><p>3.1</p></li></ul></li><li><p>4</p><ul class=\\"cherry-list__square\\"><li><p>4.2</p></li></ul></li></ul>"`;

exports[`core/hooks/list list hook 2`] = `"<ul data-lines=\\"8\\" data-sign=\\"362f490aeca95ab73384e1e30a576e1a\\" class=\\"cherry-list__default\\"><li><p>1</p><ul class=\\"cherry-list__default\\"><li><p>2</p><ul class=\\"cherry-list__default\\"><li><p>2.1</p></li><li><p>2.2</p></li></ul></li></ul></li><li><p>3<br>3.1</p></li><li><p>4</p></li><li><p>4.2</p></li></ul>"`;

exports[`core/hooks/list list hook 3`] = `"<ul data-lines=\\"1\\" data-sign=\\"95b9a2e3d59355afad427cda2937e115\\" class=\\"cherry-list__default\\"><li><p>1</p></li></ul><ol data-lines=\\"2\\" data-sign=\\"95b9a2e3d59355afad427cda2937e115\\" start=\\"1\\" class=\\"cherry-list__default\\"><li><p>test<br></p></li></ol><ul data-lines=\\"8\\" data-sign=\\"95b9a2e3d59355afad427cda2937e115\\" class=\\"cherry-list__default\\"><li><p>1.1</p><ul class=\\"cherry-list__default\\"><li><p>1.1.2</p><ul class=\\"cherry-list__default\\"><li><p>blank</p></li></ul></li><li><p>1.2</p></li></ul></li><li><p>2<br>blank</p><ul class=\\"cherry-list__default\\"><li><p>2.1</p></li><li><p>2.2</p></li></ul></li></ul><ol data-lines=\\"2\\" data-sign=\\"95b9a2e3d59355afad427cda2937e115\\" start=\\"1\\" class=\\"cherry-list__default\\"><li><p>test</p><ol start=\\"2\\" class=\\"cherry-list__default\\"><li><p>2</p></li></ol></li></ol>"`;

exports[`core/hooks/list list hook 4`] = `"<ol data-lines=\\"8\\" data-sign=\\"739b7596b094aac47153f8bcbb07e2e3\\" start=\\"1\\" class=\\"cherry-list__default\\"><li><p>test</p><ol start=\\"2\\" class=\\"cherry-list__default\\"><li><p>test</p></li></ol></li><li><p>test</p><ol start=\\"1\\" class=\\"cherry-list__cjk-ideographic\\"><li><p>test</p></li><li><p>test<br><br></p></li><li><p>test</p></li></ol></li></ol><ul data-lines=\\"1\\" data-sign=\\"739b7596b094aac47153f8bcbb07e2e3\\" class=\\"cherry-list__default\\"><li><p>test</p></li></ul>"`;

exports[`core/hooks/list list hook 5`] = `"<ul data-lines=\\"6\\" data-sign=\\"d1e2504747788dc560aedd3892b6705e\\" class=\\"cherry-list__default\\"><li class=\\"check-list-item\\"><p><span class=\\"ch-icon ch-icon-square\\"></span> checklist 1</p></li><li><p>test</p><ul class=\\"cherry-list__default\\"><li class=\\"check-list-item\\"><p><span class=\\"ch-icon ch-icon-check\\"></span> checklist 2</p></li></ul></li><li class=\\"check-list-item\\"><p><span class=\\"ch-icon ch-icon-square\\"></span> checklist 3</p></li><li><p>test</p><ul class=\\"cherry-list__default\\"><li class=\\"check-list-item\\"><p><span class=\\"ch-icon ch-icon-square\\"></span> checklist 4</p></li></ul></li></ul>"`;
